File: practice/leetcode/cpp/226_invert_binary_tree.cpp
//https://leetcode.com/problems/invert-binary-tree/description/
//Given the root of a binary tree, invert the tree, and return its root.

#include <cmath>
#include <utility>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(0), left(left), right(right) {}
};

class Solution {
public:
	TreeNode *invertTree_recursive(TreeNode *root) {
		if (root) {
			invertTree(root->right);
			invertTree(root->left);
			std::swap(root->left, root->right);
		}
		return(root);
	}
	TreeNode *invertTree_iterative(TreeNode *root) {
		stack<TreeNode *> st;
		st.push(root);

		while (!st.empty()) {
			TreeNode *tmp = st.top();
			st.pop();
			if (tmp) {
				st.push(tmp->left);
				st.push(tmp->right);
				std::swap(tmp->left, tmp->right);
			}
		}
		return(root);
	}
};

File: practice/leetcode/python/226_invert_binary_tree.py
#https://leetcode.com/problems/invert-binary-tree/description/
#Given the root of a binary tree, invert the tree, and return its root.
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root 

        root.left, root.right = root.right, root.left
        self.invertTree(root.left)
        self.invertTree(root.right)
        return root

    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root: 
            return None

        root.left, root.right = root.right, root.left
        self.invertTree(root.left)
        self.invertTree(root.right)

    def invertTree_iterative(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root

        stack = [root]
        while (len(stack)):
            node = stack.pop()
            node.left, node.right = node.right, node.left
            if node.left: 
                stack.append(node.left)
            if node.right:
                stack.append(node.right)

        return(root)

    def invertTree_iterative(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None

        st = [root]
        while (len(st)):
            node = st.pop()
            node.left, node.right = node.right, node.left
            if node.left:
                st.append(node.left)
            if node.right:
                st.append(node.right)
        return(root)
        

